improve python-agent client

sercher opened the pull request.

sercher commented:
Improved handling of python-based agent:
- send "GET /" prior to WebSocket connetion to provide the agent with endpoint address
- send "state_request" proir to "connect" to check if the user is already connected
- added random IDs to packets, where required

DevAlexey commented:
Integration test is not passing

DevAleksei commented on agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentConnection.kt:
```diff
@@ -46,7 +56,15 @@ interface Connection {
 }
 
-class AgentConnection(val myAgentUrl: String, val invite: String? = null, val userName: String = "user1", val passphrase: String = "test") : Connection {
+class AgentConnection(val myAgentUrl: String,
```
May be we should keep old API for compatibility?
Or delete invite totally?

sercher replied:
It has more to do with agent IP address, which I changed. The tests should pass when python agents are executed in a local venv rather than from the docker image.

DevAleksei replied:
I think that test should be able to pass when they are executed in any env, only IP's should change
But here i'm talking about change of parameters sequence, which will result in errors in external project without any compilation errors.

sercher replied:
> I think that test should be able to pass when they are executed in any env

Sure. It actually passes when the IP-address is specified correctly.

> But here i'm talking about change of parameters sequence, which will result in errors in external project without any compilation errors.

As agreed, I am removing "invite" parameter from the constructor and keep the method acceptInvite(). Please note there's no such thing like Old and New interface here. We're now working on the initial version.

DevAleksei replied:
If you will remove invite at all, then its totally OK, but if we were keeping it, I don't think that it is wise to swap parameters in places.

sercher replied:
apologies, that wasn't intentional.

DevAleksei commented on agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentConnection.kt:
```diff
@@ -77,29 +95,62 @@ class AgentConnection(val myAgentUrl: String, val invite: String? = null, val us
+        val uri = URI(myAgentUrl)
+        val rootPath = "http://" + uri.host + ":" + uri.port + "/"
+        val rootUrl = URL(rootPath)
+        val res = rootUrl.openConnection().getInputStream()
```
I still dont like idea, of any "magic" for agent specific implementation.
Plus, it is bad idea to keep connection opened for no real reason.

sercher replied:
Would you suggest another way of making agents to generate valid invites? Invites are core agent functionality, we can't simply omit it.

DevAleksei replied:
I think than this is agents responsibility to say on which address it can be accessed , not clients problem. Thous I would just remove all this code)
You may add this "hack" to integration test on before handler.

sercher replied:
> You may add this "hack" to integration test on before handler.

Accepted. I'll fix this.

DevAleksei commented on agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentConnection.kt:
```diff
@@ -77,29 +95,62 @@ class AgentConnection(val myAgentUrl: String, val invite: String? = null, val us
     }
 
+    override fun acceptInvite(invite: String) {
+        if (invite != null && getConnectionStatus() == ConnectionStatus.AGENT_CONNECTION_CONNECTED) {
```
I think that system should fail if we are in wrong state, it`s easier to track bugs, than hide them.

sercher replied:
Ok, this will be fixed.

DevAleksei commented on agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentConnection.kt:
```diff
@@ -214,16 +265,16 @@ class AgentWebSocketClient(serverUri: URI) : WebSocketClient(serverUri) {
 }
 
-data class WalletConnect(val name: String, val passphrase: String, @JsonProperty("@type") val type: String = CONNECT)
+data class WalletConnect(val name: String, val passphrase: String, @JsonProperty("@type") val type: String = CONNECT, val id: String? = null)
```
Why do we need Id?
And what commit of agent client you were using? 
It has changed from my version dramatically in latest commits.

sercher replied:
The id is required in certain messages by the pythonic indy-agent.
I used exactly the last commit in the branch I am sending a PR to. Does this make sense?

DevAleksei replied:
Ah, my bad, i was talking about agent service implementation.

DevAleksei commented on agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentConnection.kt:
```diff
@@ -77,29 +95,62 @@ class AgentConnection(val myAgentUrl: String, val invite: String? = null, val us
-                counterParty = IndyParty(response.their_did, invite.endpoint)
+            sendJson(StateRequest())
+            var stateResponse = waitForMessageOfType<State>(STATE_RESPONSE)
+            if(!checkState(stateResponse, userName)) {
```
i think we should keep only one approach for accepting invites.

sercher replied:
My vote would be to separate it from the constuctor. Namely, to have a method acceptInvite() in the Connection interface. What would you think?

DevAleksei replied:
Sounds good for me

sercher commented:
- get rid of an empty message sent with double-call to sendJson()
- updated pythonic indy-agent to the most recent version
- timestamp field is changed with time_sent: String in indy-agent
- added connect(), disconnect(), waitForInvitedParty() methods to the Connection interface
